### PR TITLE
Issue 263 2 d tof

### PR DESCRIPTION
**Pull request by mortenjc**
### Issue reference / description

The branch you merge from should already reference a daquiri github ticket number. You can add a descriptive title, but if an issue is referenced, you don't have to.

## Checklist for submitter

- [ ] Update docker container versions (if possible)
- [ ] Update conan package versions (if possible)
- [ ] Unit tests pass
- [ ] utils/acquire.sh runs
- [ ] utils/daquiri.sh runs
- [ ] MockProducer profile runs and produces plots (manual test)
- [ ] profiles in /data directory have been updated (if applicable)
- [ ] Request code review

### CODE REVIEW

File: source/daqlite/Custom2DTOFPlot.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    fmt::print("Supported gradients are: ");
+    for (auto &Gradient : mGradients) {
+      fmt::print("{} ", Gradient.first);
+    }
+    fmt::print("\n");
+    return mGradients.find("hot")->second;
+  }
+}
+
+std::string Custom2DTOFPlot::getNextColorGradient(std::string GradientName) {
```
Comment (JennyLouise) [inline source/daqlite/Custom2DTOFPlot.cpp:120]: Share common code with Custom2DPlot.cpp

File: source/daqlite/ESSConsumer.cpp [hunk 2]
```diff
@@ -68,6 +68,12 @@ RdKafka::KafkaConsumer *ESSConsumer::subscribeTopic() const {
 
 uint32_t ESSConsumer::processEV42Data(RdKafka::Message *Msg) {
   auto EvMsg = GetEventMessage(Msg->payload());
+  uint64_t MessageId = EvMsg->message_id();
+  if (MessageId != NextSeqNum) {
+    printf("Message sequence number error: expected %llu, got %llu\n", NextSeqNum, MessageId);
+  }
+  NextSeqNum = MessageId + 1;
+
```
Comment (JennyLouise) [inline source/daqlite/ESSConsumer.cpp:76]: Remove

File: source/daqlite/ESSConsumer.h [hunk 3]
```diff
@@ -41,13 +41,19 @@ class ESSConsumer {
...
+  std::vector<uint32_t> mTOFsPlot;
+
   uint64_t EventCount{0};
   uint64_t EventAccept{0};
   uint64_t EventDiscard{0};
 
 
 
 private:
+  uint64_t NextSeqNum{0}; // Kafka - from message_id
```
Comment (JennyLouise) [inline source/daqlite/ESSConsumer.h:56]: Remove

File: source/daqlite/MainWindow.cpp [hunk 4]
```diff
@@ -100,96 +131,91 @@ void MainWindow::handleKafkaData(int ElapsedCountMS) {
...
 
 void MainWindow::handleGradientButton() {
-  if (TOF) {
+  if (plottype == pixels) {
+    mConfig.Plot.ColorGradient =
+      Plot2DXY->getNextColorGradient(mConfig.Plot.ColorGradient);
+  } else if (plottype == tof2d) {
+    mConfig.Plot.ColorGradient =
+      PlotTOF2D->getNextColorGradient(mConfig.Plot.ColorGradient);
+      PlotTOF2D->plotDetectorImage(true);
```
Comment (JennyLouise) [inline source/daqlite/MainWindow.cpp:216]: unindent

**Review by JennyLouise — APPROVED** [review]

File: source/daqlite/Custom2DTOFPlot.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  /// \brief allocated according to config in constructor
+  #define TOF2DX 512
+  #define TOF2DY 512
+  uint32_t HistogramData2D[TOF2DX + 1][TOF2DY + 1];
+
+  /// \brief for calculating x, y, z from pixelid
+  ESSGeometry *LogicalGeometry;
+
+  // 0 = (x,y), 1 = (x,z), 2 = (y,z)
+  int mProjection{0};
```
[the remark this answers is not in the document]
Comment (mortenjc) [reply source/daqlite/Custom2DTOFPlot.h:63]: nope

File: source/daqlite/Custom2DTOFPlot.cpp [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  if (PixelIDs.size() == 0) {
+    return;
+  }
+
+  for (int i = 0; i < PixelIDs.size(); i++) {
+    if (PixelIDs[i] == 0) {
+      continue;
+    }
+    int tof  = TOFs[i];
+    int yvals = (PixelIDs[i] - 1)/32;
```
[the remark this answers is not in the document]
Comment (mortenjc) [reply source/daqlite/Custom2DTOFPlot.cpp:184]: too much work, but changed hardcoded 32 to XDim from config

**Pull request merged**